enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -1793,6 +1804,196 @@ func (p *PebbleCache) Reader(ctx context.Context, r *rspb.ResourceName, uncompre
+	cdcw.numChunks++
+
+	if cdcw.numChunks == 1 {
+		cdcw.firstChunk = make([]byte, len(chunkData))
+		copy(cdcw.firstChunk, chunkData)
+		return nil
+	}
+
+	if cdcw.numChunks == 2 {
+		if err := cdcw.writeChunkWhenMultiple(cdcw.firstChunk); err != nil {
+			return err
+		}
+	}
+	// we need to copy the data because once the chunker calls Next, chunkData
+	// will be invalidated.
+	data := make([]byte, len(chunkData))
+	copy(data, chunkData)
+	return cdcw.writeChunkWhenMultiple(data)
+}
+
+func (cdcw *cdcWriter) writeRawChunk(fileRecord *rfpb.FileRecord, key filestore.PebbleKey, chunkData []byte) error {
+	ctx := cdcw.ctx
+	p := cdcw.pc
+	inlineWriter := p.fileStorer.InlineWriter(ctx, fileRecord.GetDigest().GetSizeBytes())
+	wcm, err := p.newWrappedWriter(ctx, inlineWriter, fileRecord, key, cdcw.shouldCompress || cdcw.isCompressed)
+	if err != nil {
+		return err
+	}
+	_, err = wcm.Write(chunkData)
+	if err != nil {
+		return err
+	}
+	if err := wcm.Close(); err != nil {
+		return err
+	}
+	if err := wcm.Commit(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (cdcw *cdcWriter) writeChunkWhenSingle(chunkData []byte) error {
+	// When there is only one single chunk, we want to store the original file
+	// record with the original key instead of computed digest from the
+	// chunkData. This is because the chunkData can be compressed or encrypted,
+	// so the digest computed from it will be different from the original digest.
+	return cdcw.writeRawChunk(cdcw.fileRecord, cdcw.key, chunkData)
+}
+
+func (cdcw *cdcWriter) writeChunkWhenMultiple(chunkData []byte) error {
+	ctx := cdcw.ctx
+	p := cdcw.pc
+
+	d, err := digest.Compute(bytes.NewReader(chunkData), cdcw.fileRecord.GetDigestFunction())
+	if err != nil {
+		return err
+	}
+	r := &rspb.ResourceName{
+		Digest:         d,
+		InstanceName:   cdcw.fileRecord.GetIsolation().GetRemoteInstanceName(),
+		CacheType:      cdcw.fileRecord.GetIsolation().GetCacheType(),
+		DigestFunction: cdcw.fileRecord.GetDigestFunction(),
+		Compressor:     cdcw.fileRecord.GetCompressor(),
+	}
+	if cdcw.shouldCompress && r.Compressor == repb.Compressor_IDENTITY {
+		// we need to compress the chunk, but this data hasn't been compressed yet.
+		// so we need to set the resource name to identity to signal to the nested
+		// writer to compress it.
+		r.Compressor = repb.Compressor_IDENTITY
+	} else {
+		r.Compressor = repb.Compressor_ZSTD
+	}
+	fileRecord, err := p.makeFileRecord(ctx, r)
+
+	if err != nil {
+		return err
+	}
+	key, err := p.fileStorer.PebbleKey(fileRecord)
+	if err != nil {
+		return err
+	}
+	cdcw.writtenChunks = append(cdcw.writtenChunks, r)
+
+	cdcw.eg.Go(func() error {
+		return cdcw.writeRawChunk(fileRecord, key, chunkData)
+	})
+	return nil
+}
+
+func (cdcw *cdcWriter) Write(buf []byte) (int, error) {
+	// If the bytes being written are compressed, we decompress them in
+	// order to generate CDC chunks, then compress those chunks.
+	if cdcw.isCompressed {
```

Comment:
i wonder if we'll want to do this if the file is gonna be less than 1 chunk size? maybe some way to skip this code path?